Zilliqa/Zilliqa
# Dual PoW difficulty levels for DS committee and shards

## Note
Do not merge till kubernetes configuration is also updated 

## Description



With the merging of PoW1 and PoW2, there is currently only 1 PoW required in the protocol. This PR implements the mechanism for choosing new node into the DS committee. 

In this implementation, there will be two difficulty levels, known as DS difficulty and difficulty (for shard). Both will use the same PoW process. The former is expected to be harder than the later. 

- If a node solution meets the difficulty level, it may be chosen into shard. 
- If a node solution meets the DS difficulty level, it may be chosen to be part of the DS committee (if not, it may be chosen into shard). 
- A node may submit multiple solution*, but the solution with the hash value being the least, will be taken into account. 

\* Spammer will be blocked by reputation manager and a max submission limit.

If multiple nodes solution meet DS difficulty level, the solution with the least hash value will chosen into the DS committee. 

Todo:
- ~Node to send the difficulty level at which pow is performed~
- ~DS to verify with the correct difficulty level~
- Shard to verify ds leader + sharding structure

Other changes:
- Remove magic numbers from genesis ds block
- Rename `shouldMine` to `m_shouldMine` in libPoW
- Change `m_shouldMine` to `atomic`
- Refactor `StartPoW()` to separate the logic of computng PoW and Sending out PoW message
- Cleanup `libNode/PowProcessing.cpp`
- Remove duplicated mutex `m_mutexAllPoWs`  and unused mutex `m_mutexDSBlockConsensus` (Fix https://github.com/Zilliqa/Issues/issues/180) 

Other todos:
- Update internal testnet constants file
- Update explorer to display dual difficulty
- Dynamic DS difficulty level
- Validator for ds leader

Known issue (for core team reference):
- https://github.com/Zilliqa/Issues/issues/179
- https://github.com/Zilliqa/Issues/issues/193
- https://github.com/Zilliqa/Issues/issues/194

A follow up PR will implements multiple DS nodes joining and leaving the DS committee. 

## Review Suggestion


Do a local run and check CPU utilitization

## Status

### Implementation


### Integration Test (Core Team)



- [x] local machine test (commit: 94e02a3ae3057598287dfcb8286bb327d14b0215)
- [x] medium-scale cloud test (commit: 94e02a3ae3057598287dfcb8286bb327d14b0215)
- [x] medium-scale cloud test (commit: aaf6507558df6a6ccda27899f41413839a275502)
- [x] medium-scale cloud test (commit: 5f8df331ad9a88486da7e1f3ab964cf5eacca035)
